Core/Core/Router/CourseTabUrlInteractor.swift

```diff
@@ -0,0 +1,173 @@
+                // internal path from `routes`: "course/:courseID/front_page"
+                tabPaths.append("/\(context.pathComponent)/front_page")
+            default:
+                break
+            }
+        }
+
+        return tabPaths
+    }
+
+    /// Returns `true` if `url` is not a course tab URL OR it is but it's not in the list of enabled course tab URLs.
+    public func isAllowedUrl(_ url: URL) -> Bool {
+        // if url doesn't match ""/courses/:courseID/*" for the known courses -> it's not a tab, allow it
+        guard let context = Context(url: url), let enabledTabs = enabledTabsPerCourse[context] else {
+            return true
+        }
+
+        let relativePath = String(url.relativePath.trimmingPrefix("/api/v1"))
+
+        // if url doesn't even match known tab path formats -> it's not a tab, allow it
+        guard isKnownPathFormat(relativePath) else {
+            return true
+        }
+
+        // it's a tab, if it matches any of the enabled tabs allow it, otherwise block it
+        return enabledTabs.contains(relativePath)
+    }
+
+    // expects relative paths, with "/api/v1" already stripped
+    private func isKnownPathFormat(_ path: String) -> Bool {
+        let parts = path.split(separator: "/").map { String($0) }
+        return CourseTabFormat.allCases.contains {
+            $0.isMatch(for: parts)
+        }
+    }
+
+    private func logPathFormatIfUnknown(for tab: TabModel) {
+        guard tab.id != "home" && !isKnownPathFormat(tab.htmlUrl) else { return }
+
+        RemoteLogger.shared.logError(
+            name: "Unexpected Course Tab path format",
+            reason: "tab.id: \(tab.id), tab.html_url: \(tab.htmlUrl), baseUrl: \(Analytics.analyticsBaseUrl)"
+        )
+    }
+}
+
+/// Known course tab formats
+private enum CourseTabFormat: CaseIterable {
+    case regularTab
+    case syllabusTab
+    case frontPage
+    case externalToolTab
+
+    func isMatch(for parts: [String]) -> Bool {
+        switch self {
+
+        case .regularTab:
+            // example: "/courses/42/grades"
+            guard parts.count == 3 else { return false }
+
+            let lastPart = parts[2]
+            let internalOnlyRouteComponents = [
+                "tabs",
+                "activity_stream"
+            ]
+            // if the last part matches an internal only suffix -> it's not a tab
+            return !internalOnlyRouteComponents.contains(lastPart)
+
+        case .syllabusTab:
+            // example: "/courses/42/assignments/syllabus"
+            return parts.count == 4 && parts[3] == "syllabus"
+
+        case .frontPage:
+            // example: "/courses/42/pages/front_page"
+            return parts.count == 4 && parts[2] == "pages" && parts[3] == "front_page"
+
+        case .externalToolTab:
+            // example: "/courses/42/external_tools/1234"
+            return parts.count == 4 && parts[2] == "external_tools"
```

Comment:
Do we want to add a guard in case there's an invalid route format not to crash with the array indexing?

Reply:
Unless I'm missing something, `parts.count == 4` before the `&&` ensures that we never reach the `parts[2]` check.

But on a related note, I will update the syllabus rule to include the `assignments` part, to avoid nonexistent corner case of `external_tools/syllabus` depending simply on ordering of the cases

Reply:
True that! I was looking for a guard like on the first case.